src/inc/conpty-universal.h

```diff
@@ -30,18 +30,161 @@
 
 const unsigned int PTY_SIGNAL_RESIZE_WINDOW = 8u;
 
-HRESULT CreateConPty(const std::wstring& cmdline,       // _In_
-                     const unsigned short w,            // _In_
-                     const unsigned short h,            // _In_
-                     HANDLE* const hInput,              // _Out_
-                     HANDLE* const hOutput,             // _Out_
-                     HANDLE* const hSignal,             // _Out_
-                     PROCESS_INFORMATION* const piPty); // _Out_
+// A case-insensitive wide-character map is used to store environment variables
+// due to documented requirements:
+//
+//      "All strings in the environment block must be sorted alphabetically by name.
+//      The sort is case-insensitive, Unicode order, without regard to locale.
+//      Because the equal sign is a separator, it must not be used in the name of
+//      an environment variable."
+//      https://docs.microsoft.com/en-us/windows/desktop/ProcThread/changing-environment-variables
+//
+struct WStringCaseInsensitiveCompare
+{
+    bool operator()(const std::wstring& lhs, const std::wstring& rhs) const
+    {
+        return (::_wcsicmp(lhs.c_str(), rhs.c_str()) < 0);
+    }
+};
+
+using EnvironmentVariableMapW = std::map<std::wstring, std::wstring, WStringCaseInsensitiveCompare>;
+
+HRESULT CreateConPty(const std::wstring& cmdline,                       // _In_
+                     const unsigned short w,                            // _In_
+                     const unsigned short h,                            // _In_
+                     HANDLE* const hInput,                              // _Out_
+                     HANDLE* const hOutput,                             // _Out_
+                     HANDLE* const hSignal,                             // _Out_
+                     PROCESS_INFORMATION* const piPty,                  // _Out_
+                     const EnvironmentVariableMapW& extraEnvVars = {}); // _In_
 
 bool SignalResizeWindow(const HANDLE hSignal,
                         const unsigned short w,
                         const unsigned short h);
 
+HRESULT UpdateEnvironmentMapW(EnvironmentVariableMapW& map);
+
+HRESULT EnvironmentMapToEnvironmentStringsW(EnvironmentVariableMapW& map, std::vector<wchar_t>& newEnvVars);
+
+// Function Description:
+// - Updates an EnvironmentVariableMapW with the current process's unicode
+//   environment variables ignoring ones already set in the provided map.
+// Arguments:
+// - map: The map to populate with the current processes's environment variables.
+// Return Value:
+// - S_OK if we succeeded, or an appropriate HRESULT for failing
+__declspec(noinline) inline

```

Comment:
what is the purpose of `__declspec(noinline) inline`?

Reply:
I don't know, it's what was already there for the other functions, so I assumed it was some sort of standard practice for functions declared here.

Reply:
I think this is another question for @zadjii-msft then.

Reply:
Oh man, why that there is definitely lost to the annals of time. Someone commented on the original PR for that file that I should do that, but I can't remember exactly why. Something about the fact that we were adding a function in a header-only file, this was important. Maybe @miniksa remembers?

Reply:
The inline makes sense if you're defining the actual implementation of a function in a header since otherwise you'd get multiple linker definition errors when attempting to link together multiple objects that included the header and included the function.

The only mystery to me is the `__declspec(noinline)` combined with inline.

Reply:
`__declspec(noinline) inline` is a straight-up magic way to prevent functions from being emitted in every translation unit that includes this header. It seems like a more palatable way to say `__declspec(selectany)`...

More info [here](https://blogs.msdn.microsoft.com/freik/2005/10/26/how-to-make-a-no-inline-inline-function-and-why-you-might-want-to/).